src/modules/joinOrderOptimization.js

```diff
@@ -0,0 +1,460 @@
+        if (!shouldAddRow) break;
+      }
+      if (shouldAddRow) {
+        const newRow = {...tableARow, ...tableBRow};
+        newRows.push(newRow);
+      }
+    }
+  }
+  return {rows: newRows, cost: totalCost};
+};
+
+/**
+ * Gets the data for the given table
+ * @param tableName - the table name
+ */
+const getTableData = (tableName: string) => {
+  switch (tableName) {
+    case 'Department':
+      return addTablePrefixToData('department', department.data);
+    case 'Doctor':
+      return addTablePrefixToData('doctor', doctor.data);
+    case 'Patient':
+      return addTablePrefixToData('patient', patient.data);
+    default:
+      return [];
+  }
+};
+
+/**
+ * Adds the table prefix to the data
+ * @param tableName - the table name
+ * @param data - the data
+ */
+const addTablePrefixToData = (
+  tableName: string,
+  data: Array<{[key: string]: string}>
+) => {
+  const newData = [];
+  for (const row of data) {
+    const newRow: {[key: string]: string} = {};
+    for (const key of Object.keys(row)) {
+      newRow[`${tableName}.${key}`.toLowerCase()] = row[key];
+    }
+    newData.push(newRow);
+  }
+  return newData;
+};
+
+class JoinOrderQueueElement {
+  graph: Graph;
+  joinTables: Array<string>;
+  rows: Array<any>;
+  cost: number;
+  joinOrder: Array<string | Array<JoinCondition>>;
+
+  constructor(
+    graph: Graph,
+    joinTables: Array<string>,
+    rows: Array<any>,
+    cost: number,
+    joinOrder: Array<string | Array<JoinCondition>>
+  ) {
+    this.graph = graph;
+    this.joinTables = joinTables;
+    this.rows = rows;
+    this.cost = cost;
+    this.joinOrder = joinOrder;
+  }
+
+  getCost = (): number => {
+    return this.cost;
+  };
+
+  getSize = (): number => {
+    return this.joinTables.length;
+  };
+
+  getChildren = (): Array<JoinOrderQueueElement> => {
+    const children = [];
+    for (const table of this.joinTables) {
+      const edges = this.graph[table].edges;
+      const neighbors = Object.keys(edges);
+      for (const neighbor of neighbors) {
+        if (this.joinTables.includes(neighbor)) continue;
+        let joinConditions: Array<JoinCondition> = [];
+        for (const currentTable in this.graph[neighbor].edges) {
+          if (this.joinTables.includes(currentTable)) {
+            joinConditions = [
+              ...joinConditions,
+              ...this.graph[neighbor].edges[currentTable],
+            ];
+          }
+        }
+        const newJoinOrder = [...this.joinOrder, joinConditions, neighbor];
+        const {rows, cost} = getRowsAndCost(
+          this.rows,
+          this.cost,
+          joinConditions,
+          this.joinTables,
+          neighbor
+        );
+        children.push(
+          new JoinOrderQueueElement(
+            this.graph,
+            [...this.joinTables, neighbor],
+            rows,
+            cost,
+            newJoinOrder
+          )
+        );
+      }
+    }
+    return children;
+  };
+}
+
+class PriorityQueue {
```

Comment:
Rather than adding a priority queue implementation here, I would suggest using an existing one such as the `tinyqueue` module on npm.